File: examples/log-triage.ts
/*
 * TSN Log Triage
 *
 * Reads structured-ish log lines from stdin and prints an actionable summary.
 *
 * Compile: tsn build examples/log-triage.ts
 * Run:     ./build/log-triage < sample.log
 */

interface LogEntry {
  level: string;
  service: string;
  region: string;
  message: string;
  tagLabel: string;
}

interface ServiceStat {
  service: string;
  count: number;
}

import { createRequire } from "module";
const require = createRequire(import.meta.url);

function readStdin(): string {
  const fs = require("fs");
  return fs.readFileSync("/dev/stdin", "utf-8");
}

function fieldValue(line: string, key: string): string {
  const start: number = line.indexOf(key);
  if (start === -1) return "";
  const rest: string = line.slice(start + key.length);
  const end: number = rest.indexOf(" ");
  if (end === -1) return rest.trim();
  return rest.slice(0, end).trim();
}

function messageValue(line: string): string {
  const marker: string = " msg=";
  const start: number = line.indexOf(marker);
  if (start === -1) return line.trim();
  return line.slice(start + marker.length).trim();
}

function classifyTags(level: string, service: string, message: string): string[] {
  const tags: string[] = [];
  const lowered: string = message.toLowerCase();
  if (level === "ERROR") tags.push("page");
  if (lowered.includes("timeout") || lowered.includes("latency")) tags.push("latency");
  if (lowered.includes("payment") || service === "checkout") tags.push("payments");
  if (lowered.includes("token") || lowered.includes("auth")) tags.push("auth");
  if (lowered.includes("queue") || lowered.includes("backlog")) tags.push("queue");
  if (lowered.includes("deploy") || lowered.includes("release")) tags.push("release");
  return tags;
}

function parseEntry(line: string): LogEntry {
  const clean: string = line.trim();
  const level: string = fieldValue(clean, "level=");
  const service: string = fieldValue(clean, "service=");
  const region: string = fieldValue(clean, "region=");
  const message: string = messageValue(clean);
  const tags: string[] = classifyTags(level, service, message);
  const entry: LogEntry = {
    level: level,
    service: service,
    region: region,
    message: message,
    tagLabel: tags.join(", "),
  };
  return entry;
}

function isActionable(entry: LogEntry): boolean {
  if (entry.level === "ERROR") return true;
  if (entry.level === "WARN" && entry.tagLabel.length > 0) return true;
  if (entry.message.includes("retry storm")) return true;
  return false;
}

function bumpService(stats: ServiceStat[], service: string): ServiceStat[] {
  const next: ServiceStat[] = stats.slice(0);
  const idx: number = next.findIndex((s: ServiceStat): boolean => s.service === service);
  if (idx === -1) {
    const stat: ServiceStat = { service: service, count: 1 };
    next.push(stat);
    return next;
  }
  const current: ServiceStat = next[idx];
  const updated: ServiceStat = { service: current.service, count: current.count + 1 };
  next[idx] = updated;
  return next;
}

function parseEntries(raw: string): LogEntry[] {
  const entries: LogEntry[] = [];
  for (const line of raw.split("\n")) {
    const clean: string = line.trim();
    if (clean.length > 0) entries.push(parseEntry(clean));
  }
  return entries;
}

function printSummary(entries: LogEntry[]): void {
  const actionable: LogEntry[] = entries.filter((e: LogEntry): boolean => isActionable(e));
  const errorCount: number = entries.count((e: LogEntry): boolean => e.level === "ERROR");
  const warnCount: number = entries.count((e: LogEntry): boolean => e.level === "WARN");

  let serviceStats: ServiceStat[] = [];
  for (const entry of actionable) {
    serviceStats = bumpService(serviceStats, entry.service);
  }
  serviceStats.sort((a: ServiceStat, b: ServiceStat): number => b.count - a.count);

  const serviceNames: string[] = serviceStats.map((s: ServiceStat): string => s.service);

  console.log("=== LOG TRIAGE ===");
  console.log("Total lines: " + String(entries.length));
  console.log("Actionable: " + String(actionable.length));
  console.log("Errors: " + String(errorCount));
  console.log("Warnings: " + String(warnCount));
  console.log("");
  console.log("Hot services: " + serviceNames.join(", "));
  console.log("");

  for (const entry of actionable) {
    const label: string = "[" + entry.level + "] " + entry.service + "@" + entry.region;
    console.log(label + " :: " + entry.message + " :: " + entry.tagLabel);
  }
}

function main(): void {
  const entries: LogEntry[] = parseEntries(readStdin());
  printSummary(entries);
}

main();

File: harness/rust/log-triage.rs
// Rust port of examples/log-triage.ts — zero-dep, stdlib only.

use std::io::{self, Read, Write, BufWriter};

struct LogEntry {
    level: String,
    service: String,
    region: String,
    message: String,
    tag_label: String,
}

struct ServiceStat {
    service: String,
    count: i64,
}

fn field_value(line: &str, key: &str) -> String {
    let start = match line.find(key) {
        Some(i) => i,
        None => return String::new(),
    };
    let value_start = start + key.len();
    let rest = &line[value_start..];
    let end = rest.find(' ').unwrap_or(rest.len());
    rest[..end].trim().to_string()
}

fn message_value(line: &str) -> String {
    let marker = " msg=";
    match line.find(marker) {
        Some(i) => line[i + marker.len()..].trim().to_string(),
        None => line.trim().to_string(),
    }
}

fn classify_tags(level: &str, service: &str, message: &str) -> Vec<&'static str> {
    let mut tags = Vec::new();
    let lowered = message.to_lowercase();
    if level == "ERROR" { tags.push("page"); }
    if lowered.contains("timeout") || lowered.contains("latency") { tags.push("latency"); }
    if lowered.contains("payment") || service == "checkout" { tags.push("payments"); }
    if lowered.contains("token") || lowered.contains("auth") { tags.push("auth"); }
    if lowered.contains("queue") || lowered.contains("backlog") { tags.push("queue"); }
    if lowered.contains("deploy") || lowered.contains("release") { tags.push("release"); }
    tags
}

fn parse_entry(line: &str) -> LogEntry {
    let clean = line.trim();
    let level = field_value(clean, "level=");
    let service = field_value(clean, "service=");
    let region = field_value(clean, "region=");
    let message = message_value(clean);
    let tags = classify_tags(&level, &service, &message);
    let tag_label = tags.join(", ");
    LogEntry { level, service, region, message, tag_label }
}

fn is_actionable(e: &LogEntry) -> bool {
    if e.level == "ERROR" { return true; }
    if e.level == "WARN" && !e.tag_label.is_empty() { return true; }
    if e.message.contains("retry storm") { return true; }
    false
}

fn add_service_count(stats: &mut Vec<ServiceStat>, service: &str) {
    for s in stats.iter_mut() {
        if s.service == service {
            s.count += 1;
            return;
        }
    }
    stats.push(ServiceStat { service: service.to_string(), count: 1 });
}

fn parse_entries(raw: &str) -> Vec<LogEntry> {
    let mut entries = Vec::new();
    for line in raw.split('\n') {
        let line = line.trim();
        if !line.is_empty() {
            entries.push(parse_entry(line));
        }
    }
    entries
}

fn main() {
    let mut raw = String::new();
    io::stdin().read_to_string(&mut raw).unwrap();
    let entries = parse_entries(&raw);

    let mut actionable: Vec<&LogEntry> = Vec::new();
    let mut service_stats: Vec<ServiceStat> = Vec::new();
    let mut error_count = 0i64;
    let mut warn_count = 0i64;

    for e in &entries {
        if e.level == "ERROR" { error_count += 1; }
        if e.level == "WARN" { warn_count += 1; }
        if is_actionable(e) {
            actionable.push(e);
            add_service_count(&mut service_stats, &e.service);
        }
    }

    service_stats.sort_by(|a, b| b.count.cmp(&a.count));
    let service_names: Vec<String> = service_stats.iter().map(|s| s.service.clone()).collect();

    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    writeln!(out, "=== LOG TRIAGE ===").unwrap();
    writeln!(out, "Total lines: {}", entries.len()).unwrap();
    writeln!(out, "Actionable: {}", actionable.len()).unwrap();
    writeln!(out, "Errors: {}", error_count).unwrap();
    writeln!(out, "Warnings: {}", warn_count).unwrap();
    writeln!(out).unwrap();
    writeln!(out, "Hot services: {}", service_names.join(", ")).unwrap();
    writeln!(out).unwrap();

    for e in &actionable {
        writeln!(out, "[{}] {}@{} :: {} :: {}", e.level, e.service, e.region, e.message, e.tag_label).unwrap();
    }
}
